### PR TITLE
Expose transfer-related metrics in `Worker.get_metrics` and `WorkerMetricCollector`

### LINKED ISSUE #6892: Add metrics for worker communications
To better understand worker communications, their performance and required resources, we should track more metrics. Specifically, it would be interesting to track the number of open incoming/outgoing connections and `comm_nbytes` in `Worker.get_metrics`. This could help debug issues such as dask/distributed#6208 or https://github.com/coiled/coiled-runtime/pull/229#discussion_r946807049 and help take decisions on possible solutions for those (https://github.com/dask/distributed/issues/6208#issuecomment-1216710657)

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #6892

* Adds transfer-related metrics in `Worker.get_metrics` and `WorkerMetricCollector`
* Adds testing for `transfer_incoming_*` metrics

**Out of scope:**
* Dashboard component for `transfer_incoming_bytes` and `transfer_outgoing_bytes` to be added in a follow-up PR 

**Notes**:
* `outgoing` metrics appear to be rather awkward to test since the timing needs to be "just right".

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/http/worker/tests/test_worker_http.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ async def test_prometheus(c, s, a, b):
         assert response.headers["Content-Type"] == "text/plain; version=0.0.4"
 
         txt = response.body.decode("utf8")
-        families = {familiy.name for familiy in text_string_to_metric_families(txt)}
+        families = {family.name for family in text_string_to_metric_families(txt)}
```
Comment (hendrikmakait) [inline distributed/http/worker/tests/test_worker_http.py:28]: Driveby: Fixing typo

File: distributed/http/worker/prometheus/core.py [hunk 2]
```diff
@@ -52,6 +52,12 @@ def collect(self):
             value=self.server.latency,
         )
 
+        yield GaugeMetricFamily(
+            self.build_name("comm_reserved_bytes"),
```
Comment (hendrikmakait) [inline distributed/http/worker/prometheus/core.py:56]: `comm_reserved_bytes` feels like a more descriptive name than `comm_nbytes` in general, I'm inclined to switch that out in general.

File: distributed/http/worker/prometheus/core.py [hunk 3]
```diff
@@ -52,6 +52,12 @@ def collect(self):
...
+            "Number of bytes currently reserved for incoming/outgoing data transfers.",
```
Comment (crusaderky) [inline distributed/http/worker/prometheus/core.py:57]: Not quite right. WorkerState.comm_nbytes is exclusively about the *incoming* bytes.
The two are fundamentally different memory-wise:

- incoming bytes are going to cause a spike in unmanaged memory as soon as they arrive, and only when the transfer is complete they will transition to managed memory and become spillable. Also, no amount of pausing will make them stop coming. So they are at a very high risk of making the worker shoot past the `terminate` threshold.
- outgoing bytes are managed memory. There are two notable exceptions:
  - the key is released while the outgoing comms are in progress
  - the key is spilled out while the outgoing comms are in progress. Notice that the act of sending the key out will make it rise to the top of the SpillBuffer LRU, so this will happen only in cases of extreme spill/unspill activity.

File: distributed/http/worker/prometheus/core.py (hunk 2 above)
Comment (crusaderky) [reply distributed/http/worker/prometheus/core.py:56]: please rename to `comm_incoming_nbytes` everywhere

Comment (crusaderky) [reply distributed/http/worker/prometheus/core.py:56]: Please also update documentation at worker_state_machine.py:1143

File: distributed/worker.py [hunk 4]
```diff
@@ -974,6 +974,7 @@ async def get_metrics(self) -> dict:
                 "memory": spilled_memory,
                 "disk": spilled_disk,
             },
+            comm_reserved_bytes=self.state.comm_nbytes,
```
Comment (crusaderky) [inline distributed/worker.py:977]: Worth noting that on line 967 above,

```
in_flight=self.state.in_flight_tasks_count,
```
This is also strictly the number of *incoming* in-flight tasks.

Comment (crusaderky) [reply distributed/worker.py:977]: Outgoing traffic can be measured:

- `Worker.outgoing_current_count` for number of outgoing connections
- there is no measure for outgoing number of bytes. However it's straightforward to add it to `Worker.get_data`:
https://github.com/dask/distributed/blob/4cf9bafdd8ce6524c9e9667f544880c8e561c645/distributed/worker.py#L1688
(add to a total counter after the above line and remove the same amount in a `finally` clause)

Comment (crusaderky) [reply distributed/worker.py:977]: To clarify: it's OK to leave metering of outgoing traffic to a different PR.
However, within the scope of this PR, make it clear through naming that the "in flight tasks" and nbytes are specifically about incoming traffic.

File: distributed/http/worker/prometheus/core.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/http/worker/prometheus/core.py:56]: To clarify: Does this include renaming `WorkerState.comm_nbytes` (maybe with a deprecation warning)?

File: distributed/http/worker/prometheus/core.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/http/worker/prometheus/core.py:57]: Thanks for the clarification!

**Comment by crusaderky** [comment]
@hendrikmakait 

To recap our meeting:

- rename `WorkerState.comm_nbytes` to `comm_incoming_bytes`.
No deprecation in WorkerState needed; just change the already existing `DeprecatedWorkerStateAttribute` in Worker
- new property `WorkerState.comm_incoming_count: return len(self.in_flight_workers)`
- rename `Worker.outgoing_current_count` to `comm_outgoing_count`. Add a deprecated property (read-only).
- heartbeat will now return `{comm: {incoming_bytes: #, incoming_count: #, outgoing_bytes: #, outgoing_count: #}}`
- new attribute `Worker.outgoing_current_bytes`
- drop attributes `Worker.outgoing_count` and `Worker.incoming_count`; no deprecation needed (I doubt anybody uses them)
- update https://github.com/dask/distributed/pull/6933/ to match. Prefer "comm" prefix to "transfers" suffix everywhere.

**Comment by crusaderky** [comment]
[EDIT] let's not drop the cumulative counter; they're free

**Comment by crusaderky** [comment]
Updated naming after conversation on https://github.com/dask/distributed/pull/6933#discussion_r957430960
This supersedes my previous post.

main | new | notes
--- | --- | ---
WorkerState.comm_nbytes  |         WorkerState.transfer_incoming_bytes_current
(new property)            |        WorkerState.transfer_incoming_count_current | `len(self.in_flight_workers)`
WorkerState.total_out_connections| WorkerState.transfer_incoming_count_max
WorkerState.comm_threshold_bytes|  WorkerState.transfer_incoming_bytes_max
WorkerState.target_message_size |  (do not change? please discuss)
Worker.outgoing_current_count  |   Worker.transfer_outgoing_count_current
(new attribute)                |   Worker.transfer_outgoing_bytes_current
(new attribute)               |    Worker.transfer_outgoing_bytes_cumulative
Worker.incoming_count          |   WorkerState.transfer_incoming_count_cumulative | Moved to WorkerState
Worker.outgoing_count         |   Worker.transfer_outgoing_count_cumulative
Worker.max_connections			|  Worker.transfer_outgoing_count_max

Heartbeat will return
```yaml
transfer:
    incoming_bytes_current: ...
    etc.
```

**Comment by crusaderky** [comment]
Note that the above list does not include
- transfer_incoming_bytes_cumulative
- transfer_outgoing_bytes_cumulative

They would be new attributes, but I'd rather leave them for a future PR if and when we realize they're actually useful.
Their meaning would also be non-trivial - are they
- the bytes requested, or
- the bytes actually received, or
- the bytes actually received so far + incoming_current + an adjustment of actual vs. expected at the end of every transfer?

**Comment by hendrikmakait** [comment]
I will `WorkerState.comm_threshold_bytes` to `WorkerState.transfer_incoming_size_throttle_threshold`, as it is used as a _minimum_ size of individual transfers to be affected by throttling.

**Comment by crusaderky** [comment]
> I will `WorkerState.comm_threshold_bytes` to `WorkerState.transfer_incoming_size_throttle_threshold`, as it is used as a _minimum_ size of individual transfers to be affected by throttling.

did you mean `transfer_incoming_bytes_throttle_threshold`? +1 from me

**Comment by hendrikmakait** [comment]
> did you mean `transfer_incoming_bytes_throttle_threshold`? +1 from me

I wanted something that didn't sound like it would be related to `transfer_incoming_bytes` since it will be on individual transfers, not the summed value of `transfer_incoming_bytes`, that's why I went with `size` instead of `bytes. No strong opinions on the eventual name though.

**Comment by hendrikmakait** [comment]
Regarding `WorkerState.target_message_size`: We _could_ rename this to `WorkerState.transfer_(incoming_)message_target_bytes` but I don't have strong opinions on that particular variable.

**Comment by crusaderky** [comment]
> Regarding `WorkerState.target_message_size`: We _could_ rename this to `WorkerState.transfer_(incoming_)message_target_bytes` but I don't have strong opinions on that particular variable.

this makes sense

**Comment by hendrikmakait** [comment]
> > Regarding `WorkerState.target_message_size`: We _could_ rename this to `WorkerState.transfer_(incoming_)message_target_bytes` but I don't have strong opinions on that particular variable.
> 
> this makes sense

Then let's go with `WorkerState.transfer_message_target_bytes` for brevity as there is currently no conflicting attribute for outgoing messages.

**Comment by hendrikmakait** [comment]
Waiting for #6936 to get merged before giving this a once-over with fresh names.

**Comment by fjetter** [comment]
> Waiting for https://github.com/dask/distributed/pull/6936 to get merged before giving this a once-over with fresh names.

Damn deadlocks everywhere... 


you're self referencing this issue ;)

**Comment by hendrikmakait** [comment]
> Damn deadlocks everywhere...
>
> you're self referencing this issue ;)

I guess we need to start linting comments, I don't see another way.

File: distributed/http/worker/prometheus/core.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/http/worker/prometheus/core.py]: I've overhauled this entire section after #6933

File: distributed/worker.py (hunk 4 above)
Comment (hendrikmakait) [reply distributed/worker.py]: Added `transfer_outgoing_bytes`

Comment (hendrikmakait) [reply distributed/worker.py]: Regarding `in_flight`: Can we safely rename this?

File: distributed/worker.py [hunk 5]
```diff
@@ -385,8 +385,10 @@ class Worker(BaseWorker, ServerNode):
     profile_history: deque[tuple[float, dict[str, Any]]]
     transfer_incoming_log: deque[dict[str, Any]]
     transfer_outgoing_log: deque[dict[str, Any]]
-    #: Number of total data transfers to other workers.
+    #: Total number of data transfers to other workers since the worker was started.
     transfer_outgoing_count_total: int
+    #: Total size of open data transfers to other worker.
```
Comment (crusaderky) [inline distributed/worker.py:390]:
```suggestion
    #: Current total size of open data transfers to other workers
```

File: distributed/worker.py [hunk 6]
```diff
@@ -385,8 +385,10 @@ class Worker(BaseWorker, ServerNode):
...
+    transfer_outgoing_bytes: int
     #: Number of open data transfers to other workers.
```
Comment (crusaderky) [inline distributed/worker.py:392]:
```suggestion
    #: Current number of open data transfers to other workers
```

File: distributed/worker.py [hunk 7]
```diff
@@ -1696,7 +1708,11 @@ async def get_data(
                     )
 
         msg = {"status": "OK", "data": {k: to_serialize(v) for k, v in data.items()}}
-        nbytes = {k: self.state.tasks[k].nbytes for k in data if k in self.state.tasks}
+        bytes_per_task = {
+            k: self.state.tasks[k].nbytes for k in data if k in self.state.tasks
+        }
+        total_bytes = sum(filter(None, bytes_per_task.values()))
```
Comment (crusaderky) [inline distributed/worker.py:1714]:
```suggestion
        # Note: `if k in self.data` above guarantees that 
        # k is in self.state.tasks too and that nbytes is non-None
        bytes_per_task = {k: self.state.tasks[k].nbytes for k in data}
        total_bytes = sum(bytes_per_task.values())
```

**Review by crusaderky — CHANGES_REQUESTED** [review]

**Review by crusaderky — APPROVED** [review]

**Comment by crusaderky** [comment]
ty!
